### ISSUE TITLE
Implementing A Unified Lottery Ticket Hypothesis for Graph Neural Networks

### ISSUE THREAD

**Issue by PrathamSoni** [pyg-team/pytorch_geometric#6935]
### 🚀 The feature, motivation and pitch

ULTH (https://arxiv.org/pdf/2102.06790.pdf) pruning on graphs allows for unified learned masks on both the graph and GNN weights that extend the notion of graph pruning. This allows significant improvements in computational efficiency without dropoffs in performance. 

We implement the ULTH GLT algorithm with three classes and define the interface as follows:
1) GLTMask: Generates an internal dictionary of masks that will be applied to the model. All named parameters will be added to the mask other than those specified by _ignore_keys_.
2) PrunedModel: Serves as a wrapper for PyG modules to inject the appropriate masks defined by GLTMask for adjacency data and model weights. 
3) GLTSearch: Performs the ULTH search algorithm as presented in the paper. First trains the masks using UGS loss (regularizes mask norms, each regularization param is individually set). Then 0s the mask to enforce sparsity to hit the desired level as specified by pruning rate arguments (again can have different rates for adjacency vs model masks). Final fine-tune of the resultant masked graph + model.

We also provide a testing script to compare results against the paper on small datasets.

### Additional context

CC: @ydalmia @nathankim7

**Comment by rusty1s** [comment]
Thanks @PrathamSoni. I think this reads all good, let me give feedback directly in the PR.
